### DISCUSSION TITLE
Wipe the content of my tables?

### DISCUSSION THREAD

**Discussion by madsane29** [simolus3/drift#2088] [Q&A] [answered]
Sometimes I want to wipe the data of the database. Before I had a `clearTable()` method in every DAO for every table and called it one by one. This meant 30+ `dao.clearTable()` call. I wanted to reduce the code + sometimes it happened that I forgot to change the `wipe data` method and add the new `dao.clearTable()` to it. So I wondered if I could just list the tables, get their names and use the `delete from $tablename` sql to wipe the content.

The problem is when I list the tables, it also lists 2 tables that are created by the sqlite. I rather not wipe the data of it.

I solved it by filtering the list, but if in the future an sqlite version decides to create an extra table, this kind of method would wipe the data of it...

```
Future<void> clearDb() async {
    List<QueryRow> result = await _getExistingTableRows();
    List<String> tableNameList = _queryRowsToTableNames(result);
    for (String tableName in tableNameList) {
      await customStatement('PRAGMA foreign_keys = OFF');
      await customStatement("delete from $tableName;");
      await customStatement('PRAGMA foreign_keys = ON');
    }
  }

  List<String> _queryRowsToTableNames(List<QueryRow> rows) => rows.map((row) => row.data["name"] as String).toList();

  Future<List<QueryRow>> _getExistingTableRows() async {
    List<String> sqliteTableNames = ["android_metadata", "sqlite_sequence"];
    String selectTablesSql = "select * from sqlite_master where type = \'table\';";
    final result = await customWriteReturning(selectTablesSql);
    result.removeWhere((row) => sqliteTableNames.contains(row.data["name"]));
    return result;
  }
```

I wonder if it is possible to make it safe... Like do we have in sqlite some kind of flags that says "this is created by sqlite"/"this is created by the developer"...?

**Comment by simolus3** [comment, marked as answer]
As drift of course needs to know about all your tables, you could do this:

```dart
Future<void> clearDb() async{
  await customStatement('pragma foreign_keys = off;');
  try {
    for (final table in allTables) {
      await delete(table).go(); // Delete everything from this table
    }
  } finally {
    await customStatement('pragma foreign_keys = on;');
  }
}
```

I think this is a lot safer since you only delete the tables that were definitely created by you.

**Reply by madsane29** [reply]
Seems logical to use the `tables` that I have in `@DriftDatabase(tables: [...], daos: [...])` but I can't really achieve to access it.

![image](https://user-images.githubusercontent.com/40585482/193459375-c4424fe8-7bce-4943-b0f2-06acca4e3b28.png)


Looking at `_$AppDatabase` in my `database.g.dart` file I found a method called `allTables`: 

![image](https://user-images.githubusercontent.com/40585482/193459415-15347500-d2b9-489b-91df-8316d4a9f179.png)

I wonder if you meant this?

Thanks in advance.

**Reply by simolus3** [reply]
Ah yes, I meant `allTables`.
